### PR TITLE
[SPARK-24653][tests] Avoid cross-job pollution in TestUtils / SpillListener.

### PR DESCRIPTION
**Pull request by vanzin**
There is a narrow race in this code that is caused when the code being
run in assertSpilled / assertNotSpilled runs more than a single job.

SpillListener assumed that only a single job was run, and so would only
block waiting for that single job to finish when `numSpilledStages` was
called. But some tests (like SQL tests that call `checkAnswer`) run more
than one job, and so that wait was basically a no-op.

This could cause the next test to install a listener to receive events
from the previous job. Which could cause test failures in certain cases.

The change fixes that race, and also uninstalls listeners after the
test runs, so they don't accumulate when the SparkContext is shared
among multiple tests.

### CODE REVIEW

**Comment by jiangxb1987** [comment]
Seems the JIRA number is not related?

File: core/src/main/scala/org/apache/spark/TestUtils.scala [hunk 1]
```diff
@@ -173,21 +173,23 @@ private[spark] object TestUtils {
    * Run some code involving jobs submitted to the given context and assert that the jobs spilled.
    */
   def assertSpilled[T](sc: SparkContext, identifier: String)(body: => T): Unit = {
-    val spillListener = new SpillListener
-    sc.addSparkListener(spillListener)
-    body
-    assert(spillListener.numSpilledStages > 0, s"expected $identifier to spill, but did not")
+    withListener(sc, new SpillListener) { listener =>
+      val ret = body
+      assert(listener.numSpilledStages > 0, s"expected $identifier to spill, but did not")
```
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/TestUtils.scala]: Maybe I'm missing something obvious, but shouldn't we run the `body` before assert?

File: core/src/main/scala/org/apache/spark/TestUtils.scala [hunk 2]
```diff
@@ -173,21 +173,23 @@ private[spark] object TestUtils {
    * Run some code involving jobs submitted to the given context and assert that the jobs spilled.
    */
   def assertSpilled[T](sc: SparkContext, identifier: String)(body: => T): Unit = {
-    val spillListener = new SpillListener
-    sc.addSparkListener(spillListener)
-    body
-    assert(spillListener.numSpilledStages > 0, s"expected $identifier to spill, but did not")
+    withListener(sc, new SpillListener) { listener =>
+      val ret = body
```
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/TestUtils.scala]: Maybe I'm missing something obvious, but why shall we need the return value here?

**Comment by vanzin** [comment]
Oops, no idea how I got the wrong bug.

File: core/src/main/scala/org/apache/spark/TestUtils.scala (hunk 2 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/TestUtils.scala]: I saw the return type in the closure, but the method itself returns `Unit`, so all that can be cleaned up.

**Comment by squito** [comment]
lgtm

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by squito** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/TestUtils.scala [hunk 3]
```diff
@@ -233,6 +233,21 @@ private[spark] object TestUtils {
     }
   }
 
+  /**
+   * Runs some code with the given listener installed in the SparkContext. After the code runs,
+   * this method will wait until all events posted to the listener bus are processed, and then
+   * remove the listener from the bus.
+   */
+  def withListener[L <: SparkListener](sc: SparkContext, listener: L) (body: L => Unit): Unit = {
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/TestUtils.scala]: private? hardly matters.

**Review by srowen — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Merged to master

**Pull request closed without merging**